src/pgtz/src/timezone.rs

```diff
@@ -0,0 +1,497 @@
+
+        if actual.len() != format.len() {
+            continue;
+        }
+
+        for (i, (atok, etok)) in actual.zip(format).enumerate() {
+            match (atok, etok) {
+                (Colon, Colon) | (Plus, Plus) => { /* Matching punctuation */ }
+                (Dash, Dash) => {
+                    is_positive = false;
+                }
+                (Num(val, _), Num(_, _)) => {
+                    let val = *val;
+                    match (hour_offset, minute_offset, second_offset) {
+                        (None, None, None) => {
+                            // Postgres allows timezones in the range -15:59:59..15:59:59
+                            if val <= 15 {
+                                hour_offset = Some(i32::try_from(val).expect(
+                                    "number between 0 and 15 should fit in signed 32-bit integer",
+                                ));
+                            } else {
+                                return Err(format!(
+                                    "Invalid timezone string ({}): timezone hour invalid {}",
+                                    value, val
+                                ));
+                            }
+                        }
+                        (Some(_), None, None) => {
+                            if val < 60 {
+                                minute_offset = Some(i32::try_from(val).expect(
+                                    "number between 0 and 59 should fit in signed 32-bit integer",
+                                ));
+                            } else {
+                                return Err(format!(
+                                    "Invalid timezone string ({}): timezone minute invalid {}",
+                                    value, val
+                                ));
+                            }
+                        }
+                        (Some(_), Some(_), None) => {
+                            if val < 60 {
+                                second_offset = Some(i32::try_from(val).expect(
+                                    "number between 0 and 59 should fit in signed 32-bit integer",
+                                ));
+                            } else {
+                                return Err(format!(
+                                    "Invalid timezone string ({}): timezone second invalid {}",
+                                    value, val
+                                ));
+                            }
+                        }
+                        // We've already seen an hour a minute and a second so we should
+                        // never see another number
+                        (Some(_), Some(_), Some(_)) => {
+                            return Err(format!(
+                                "Invalid timezone string ({}): invalid value {} at token index {}",
+                                value, val, i
+                            ))
```

Comment:
Untested.

The comment above that "we should never see another number" surprised me a bit because the timestamp could contain milliseconds. I tested that timestamps with ms are working.

Reply:
Code movement.